### PR TITLE
Harmony 1460 - Fix issues when user_work ready_count is greater than 0 but there are no ready work items

### PR DESCRIPTION
**Pull request by vinnyinverso**
## Jira Issue ID
HARMONY-1460

## Description
Added logic so that when the ready_count is out of sync (does not match the actual number of ready work items) for a user_work table record, a recalculation is triggered.
I thought that it might be a good idea to also recalculate the running count if the ready count is out of sync, to be on the safe side, so it does both when running_count is out of sync.

## Local Test Steps
Edit one of the records in the user_work table that has ready_count = 0 to something greater than zero. The next time that a worker polls for work for that particular service and job the count should get set back to zero. Also, if you set your log level (http://localhost:3000/admin/configuration/log-level?level=warn) you should see messages regarding the recalculation:

> user_work is out of sync for user vinverso and job 5ca6b5f9-44eb-42b2-9ab0-c97ea1c2e731, could not find ready work item
> recalculating ready and running counts for job 5ca6b5f9-44eb-42b2-9ab0-c97ea1c2e731

## PR Acceptance Checklist
* [x] Acceptance criteria met
* [x] Tests added/updated (if needed) and passing
* [ ] Documentation updated (if needed)

### CODE REVIEW

**Comment by chris-durbin** [comment]
> @chris-durbin would you mind weighing in on this thought when you have a chance?
> 
> > I thought that it might be a good idea to also recalculate the running count if the ready count is out of sync, to be on the safe side, so it does both when running_count is out of sync.

Agree - if one of the counts is out of sync it is likely that both are and we should recalculate both.

**Comment by vinnyinverso** [comment]
Should recalculation be scoped to the service or the whole job?

File: app/models/user-work.ts [hunk 1]
```diff
@@ -211,11 +229,42 @@ export async function recalculateReadyCount(tx: Transaction, jobID: string): Pro
...
+  await recalculateCount(tx, jobID, 'ready');
+}
+
+/**
+ * Sets the running_count to the appropriate value for each row in the user_work table for the
+ * provided jobID.
+ * @param tx - The database transaction
+ * @param jobID - The job ID
+ */
+export async function recalculateRunningCount(tx: Transaction, jobID: string): Promise<void> {
```
Comment (ygliuvt) [inline app/models/user-work.ts:264]: Remove this function. It is not called anywhere.

**Review by ygliuvt — APPROVED** [review]

**Review by chris-durbin — APPROVED** [review]
Tested out in sandbox by changing the counts manually in the database and setting things to 'running'. Verified the messages were logged and the counts were recalculated successfully. No deadlocks encountered.

**Comment by chris-durbin** [comment]
> Tested out in sandbox by changing the counts manually in the database and setting things to 'running'. Verified the messages were logged and the counts were recalculated successfully. No deadlocks encountered.

Spoke too soon - there was a deadlock, but it looks to be between the work failer and a work item update, so this change still looks fine:

```
{"application":"work-failer","env_name":"harmony-cdd-sandbox","level":"info","message":"Work failer processing started.","timestamp":"2023-05-17T18:21:24.824Z"}
{"env_name":"harmony-cdd-sandbox","level":"debug","message":"Threshold is 88748","timestamp":"2023-05-17T18:21:24.843Z"}
{"application":"work-failer","env_name":"harmony-cdd-sandbox","level":"error","message":"Work failer encountered an unexpected error","timestamp":"2023-05-17T18:21:24.843Z"}
{"application":"work-failer","env_name":"harmony-cdd-sandbox","level":"error","message":"Cannot read properties of null (reading 'valueOf')","stack":"TypeError: Cannot read properties of null (reading 'valueOf')\n    at /home/ec2-user/harmony/app/workers/work-failer.ts:76:57\n    at Array.filter (<anonymous>)\n    at WorkFailer.getExpiredWorkItems (/home/ec2-user/harmony/app/workers/work-failer.ts:72:36)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async WorkFailer.handleWorkItemUpdates (/home/ec2-user/harmony/app/workers/work-failer.ts:101:65)\n    at async WorkFailer.start (/home/ec2-user/harmony/app/workers/work-failer.ts:159:9)","timestamp":"2023-05-17T18:21:24.845Z"}
{"application":"frontend","env_name":"harmony-cdd-sandbox","level":"error","message":"Work item update failed for work item 1065 and status failed","requestId":"9090a5d9-e222-4441-b440-cc02ba12be32","requestUrl":"/workflow-ui/b9956c60-0a44-4207-9d28-75dc17d51856/1065/retry","timestamp":"2023-05-17T18:21:25.055Z","user":"cdurbin","workItemId":1065}
{"application":"frontend","code":"40P01","detail":"Process 7542 waits for ShareLock on transaction 24707; blocked by process 5532.\nProcess 5532 waits for ShareLock on transaction 24710; blocked by process 7542.","env_name":"harmony-cdd-sandbox","file":"deadlock.c","hint":"See server log for query details.","length":327,"level":"error","line":"1146","message":"update \"user_work\" set \"ready_count\" = \"ready_count\" + $1, \"running_count\" = \"running_count\" - $2 where \"job_id\" = $3 and \"service_id\" = $4 - deadlock detected","name":"error","requestId":"9090a5d9-e222-4441-b440-cc02ba12be32","requestUrl":"/workflow-ui/b9956c60-0a44-4207-9d28-75dc17d51856/1065/retry","routine":"DeadLockReport","severity":"ERROR","stack":"error: update \"user_work\" set \"ready_count\" = \"ready_count\" + $1, \"running_count\" = \"running_count\" - $2 where \"job_id\" = $3 and \"service_id\" = $4 - deadlock detected\n    at Parser.parseErrorMessage (/home/ec2-user/harmony/node_modules/pg-protocol/src/parser.ts:369:69)\n    at Parser.handlePacket (/home/ec2-user/harmony/node_modules/pg-protocol/src/parser.ts:188:21)\n    at Parser.parse (/home/ec2-user/harmony/node_modules/pg-protocol/src/parser.ts:103:30)\n    at Socket.<anonymous> (/home/ec2-user/harmony/node_modules/pg-protocol/src/index.ts:7:48)\n    at Socket.emit (node:events:513:28)\n    at Socket.emit (node:domain:489:12)\n    at addChunk (node:internal/streams/readable:315:12)\n    at readableAddChunk (node:internal/streams/readable:289:9)\n    at Socket.Readable.push (node:internal/streams/readable:228:10)\n    at TCP.onStreamRead (node:internal/stream_base_commons:190:23)","timestamp":"2023-05-17T18:21:25.056Z","user":"cdurbin","where":"while locking tuple (1,10) in relation \"user_work\"","workItemId":1065}
```

**Pull request merged**